packages/pools/src/routes.ts

```diff
@@ -0,0 +1,387 @@
+            ) {
+              usedFirstPoolMap.set(
+                highestNormalizedLiquidityFirstPool.id,
+                true
+              );
+              usedSecondPoolMap.set(
+                highestNormalizedLiquiditySecondPool.id,
+                true
+              );
+              filteredRoutePaths.push({
+                pools: [
+                  highestNormalizedLiquidityFirstPool,
+                  highestNormalizedLiquiditySecondPool,
+                ],
+                tokenOutDenoms: [intermediateDenom, tokenOutDenom],
+                tokenInDenom,
+              });
+            }
+          }
+        }
+      }
+    );
+
+    filteredRoutePaths = filteredRoutePaths.sort((path1, path2) => {
+      // Priority is given to direct swap.
+      // For direct swap, sort by normalized liquidity.
+      // In case of multihop swap, sort by first normalized liquidity.
+
+      const path1IsDirect = path1.pools.length === 1;
+      const path2IsDirect = path2.pools.length === 1;
+      if (!path1IsDirect || !path2IsDirect) {
+        return path1IsDirect ? -1 : 1;
+      }
+
+      const path1NormalizedLiquidity = path1.pools[0].getNormalizedLiquidity(
+        tokenInDenom,
+        path1.tokenOutDenoms[0]
+      );
+      const path2NormalizedLiquidity = path2.pools[0].getNormalizedLiquidity(
+        tokenInDenom,
+        path2.tokenOutDenoms[0]
+      );
+
+      return path1NormalizedLiquidity.gte(path2NormalizedLiquidity) ? -1 : 1;
+    });
+
+    this.candidatePathsCache.set(cacheKey, filteredRoutePaths);
+
+    return filteredRoutePaths;
+  }
+
+  getOptimizedRoutesByTokenIn(
+    tokenIn: {
+      denom: string;
+      amount: Int;
+    },
+    tokenOutDenom: string,
+    maxPools: number
+  ): RoutePathWithAmount[] {
+    if (!tokenIn.amount.isPositive()) {
+      throw new Error("Token in amount is zero or negative");
+    }
+
+    let paths = this.getCandidatePaths(tokenIn.denom, tokenOutDenom, true);
+    if (paths.length === 0) {
+      throw new NoPoolsError();
+    }
+
+    paths = paths.slice(0, maxPools);
+
+    const initialSwapAmounts: Int[] = [];
+    let totalLimitAmount = new Int(0);
+    for (const path of paths) {
+      const limitAmount = path.pools[0].getLimitAmountByTokenIn(tokenIn.denom);
+
+      totalLimitAmount = totalLimitAmount.add(limitAmount);
+
+      if (totalLimitAmount.lt(tokenIn.amount)) {
+        initialSwapAmounts.push(limitAmount);
+      } else {
+        let sumInitialSwapAmounts = new Int(0);
+        for (const initialSwapAmount of initialSwapAmounts) {
+          sumInitialSwapAmounts = sumInitialSwapAmounts.add(initialSwapAmount);
+        }
+
+        const diff = tokenIn.amount.sub(sumInitialSwapAmounts);
+        initialSwapAmounts.push(diff);
+
+        break;
+      }
+    }
+
+    // No enough liquidity
+    if (totalLimitAmount.lt(tokenIn.amount)) {
+      throw new NotEnoughLiquidityError();
+    }
+
+    // TODO: ...
```

Comment:
Please remove or clarify this comment

Reply:
This was to be added / modified after this PR. Keeping this for now.